### PR TITLE
Add numerical equivalence check to GraphOptz.optimizeBatchNormAfterCo…

### PR DESCRIPTION
**Pull request by mvsampath**
**Summary:**
Glow performs several optimizations on the graph that is passed to it as input.
GraphOptzTest has a test case named optimizeBatchNormAfterConvWithPred  which tests one such optimization. The task is to add a numerical equivalence check to this test case which checks that both the original and optimized graph return the same result for the same input.
Documentation:

**Test Plan:**
ninja all && tests/GraphOptzTest
[----------] Global test environment tear-down
[==========] 120 tests from 4 test suites ran. (823 ms total)
[  PASSED  ] 120 tests.

### CODE REVIEW

File: tests/unittests/GraphOptzTest.cpp [hunk 1]
```diff
@@ -357,7 +357,7 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvWithReshapeConst) {
 /// Check that the batch normalization optimization is
 /// not blocked by predicates and that it preserves them.
 TEST_F(GraphOptz, optimizeBatchNormAfterConvWithPred) {
-  Node *A =
+  Placeholder *A =
```
Comment (opti-mix) [inline tests/unittests/GraphOptzTest.cpp]: You use Placeholder here and Node for the next line, which is inconsistent. Either use `Placeholder *` for both or `auto *` for both.

File: tests/unittests/GraphOptzTest.cpp [hunk 2]
```diff
@@ -373,17 +373,21 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvWithPred) {
...
   ::glow::convertPlaceholdersToConstants(F_, bindings_, {});
-  ::glow::optimize(F_, CompilationMode::Infer);
-  EXPECT_EQ(F_->getNodes().size(), 2);
+  optimizedF_ = optimizeFunction(F_);
 
-  ASSERT_EQ(A->getNumUsers(), 1);
-  Node *newCV = A->getUsers().begin()->getUser();
+  EXPECT_EQ(optimizedF_->getNodes().size(), 2);
+
+  ASSERT_EQ(A->getNumUsers(), 2);
```
Comment (opti-mix) [inline tests/unittests/GraphOptzTest.cpp]: This condition was comparing with 1 in the old version, now it is 2. Why?

Comment (mvsampath) [reply tests/unittests/GraphOptzTest.cpp]: With new changes, when optimizedF_ = optimizeFunction(F_); is executed, number of users associated with Node A changes from 1 to 2.

static Function *optimizeFunction(Function *F) {
  auto *G = F->clone(F->getName().str() + "_optimized");
  ::glow::optimize(G, CompilationMode::Infer);
  return G;
}

File: tests/unittests/GraphOptzTest.cpp [hunk 3]
```diff
@@ -488,6 +492,7 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvButVarReused) {
   EXPECT_EQ(varFilter, CV->getFilter());
   EXPECT_TRUE(llvm::isa<BatchNormalizationNode>(optimizedF_ret->getInput()));
 
+
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]: Why did you add this extra blank line?

File: tests/unittests/GraphOptzTest.cpp [hunk 4]
```diff
@@ -373,17 +373,21 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvWithPred) {
   EXPECT_EQ(F_->getNodes().size(), 3);
 
   ::glow::convertPlaceholdersToConstants(F_, bindings_, {});
-  ::glow::optimize(F_, CompilationMode::Infer);
-  EXPECT_EQ(F_->getNodes().size(), 2);
+  optimizedF_ = optimizeFunction(F_);
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]: I think you *might* have to clone the function manually and then call `convertPlaceholdersToConstants` and `optimize` on the clone. Based on a cursory examination of the optimization I think it should be fine, but it's a good idea to check the dot graphs of `F_` and `optimizedF_` to confirm.

**Comment by SplitInfinity** [comment]
Also, please squash all of your commits into one before landing.

File: tests/unittests/GraphOptzTest.cpp (hunk 3 above)
Comment (mvsampath) [reply tests/unittests/GraphOptzTest.cpp]: My bad. I thought it was added by a ./utils/format.sh.
Will fix.

File: tests/unittests/GraphOptzTest.cpp (hunk 4 above)
Comment (mvsampath) [reply tests/unittests/GraphOptzTest.cpp]: For some reason I am not able to upload the file. May be due to company firewall. Pasting them here.

F_.dot:
digraph DAG {
	rankdir=TB;
conv[
	label = "{{<Input>Input|<Filter>Filter|<Bias>Bias}|{Convolution\lname : conv\lPredicate : Yes\lInput : float\<1 x 10 x 20 x 3\>\lFilter : float\<16 x 5 x 5 x 3\>\lBias : float\<16\>\lKernels : [5, 5]\lStrides : [1, 1]\lPads : [2, 2, 2, 2]\lGroup : 1\lDilation : 1\lLayout : NHWC\lFusedActivation : \lusers : 1\lResult : float\<1 x 10 x 20 x 16\>\l}|{<Result>Result}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Thistle1
penwidth = 2];

predicate[
	label = "{{Placeholder\lname : \"predicate\"\llayout : *\loutput : float\<1\>\lusers : 1\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

A[
	label = "{{Placeholder\lname : \"A\"\llayout : *\loutput : float\<1 x 10 x 20 x 3\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

filter__1[
	label = "{{Constant\lname : \"filter__1\"\llayout : *\loutput : float\<16 x 5 x 5 x 3\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

bias__1[
	label = "{{Constant\lname : \"bias__1\"\llayout : *\loutput : float\<16\>\lusers : 2\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

batch[
	label = "{{<Input>Input|<Scale>Scale|<Bias>Bias|<Mean>Mean|<Var>Var}|{BatchNormalization\lname : batch\lPredicate : Yes\lInput : float\<1 x 10 x 20 x 16\>\lScale : float\<16\>\lBias : float\<16\>\lMean : float\<16\>\lVar : float\<16\>\lChannelIdx : 3\lEpsilon : 1.000000e-04\lMomentum : 9.000000e-01\lusers : 1\lResult : float\<1 x 10 x 20 x 16\>\l}|{<Result>Result}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=PaleTurquoise2
penwidth = 2];

predicate__1[
	label = "{{Placeholder\lname : \"predicate__1\"\llayout : *\loutput : float\<1\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

scale__1[
	label = "{{Constant\lname : \"scale__1\"\llayout : *\loutput : float\<16\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

mean__1[
	label = "{{Constant\lname : \"mean__1\"\llayout : *\loutput : float\<16\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

variance__1[
	label = "{{Constant\lname : \"variance__1\"\llayout : *\loutput : float\<16\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

ret_save[
	label = "{{<Input>Input|<Output>Output}|{Save\lname : ret_save\lInput : float\<1 x 10 x 20 x 16\>\lOutput : float\<1 x 10 x 20 x 16\>\lusers : 0\l}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=PowderBlue
penwidth = 2];

ret[
	label = "{{Placeholder\lname : \"ret\"\llayout : *\loutput : float\<1 x 10 x 20 x 16\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

ret:Output -> ret_save:Output [dir="both"];
variance__1:Output -> batch:Var;
filter__1:Output -> conv:Filter;
mean__1:Output -> batch:Mean;
A:Output -> conv:Input;
scale__1:Output -> batch:Scale;
bias__1:Output -> batch:Bias;
conv:Result -> batch:Input;
predicate__1:Output -> batch:w;
batch:Result -> ret_save:Input;
predicate:Output -> conv:w;
bias__1:Output -> conv:Bias;
}

optimizedF_.dot:
digraph DAG {
	rankdir=TB;
conv[
	label = "{{<Input>Input|<Filter>Filter|<Bias>Bias}|{Convolution\lname : conv\lPredicate : Yes\lInput : float\<1 x 10 x 20 x 3\>\lFilter : float\<16 x 5 x 5 x 3\>\lBias : float\<16\>\lKernels : [5, 5]\lStrides : [1, 1]\lPads : [2, 2, 2, 2]\lGroup : 1\lDilation : 1\lLayout : NHWC\lFusedActivation : \lusers : 1\lResult : float\<1 x 10 x 20 x 16\>\l}|{<Result>Result}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Thistle1
penwidth = 2];

predicate__1[
	label = "{{Placeholder\lname : \"predicate__1\"\llayout : *\loutput : float\<1\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

A[
	label = "{{Placeholder\lname : \"A\"\llayout : *\loutput : float\<1 x 10 x 20 x 3\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

filter__2[
	label = "{{Constant\lname : \"filter__2\"\llayout : *\loutput : float\<16 x 5 x 5 x 3\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

bias__2[
	label = "{{Constant\lname : \"bias__2\"\llayout : *\loutput : float\<16\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

ret_save[
	label = "{{<Input>Input|<Output>Output}|{Save\lname : ret_save\lInput : float\<1 x 10 x 20 x 16\>\lOutput : float\<1 x 10 x 20 x 16\>\lusers : 0\l}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=PowderBlue
penwidth = 2];

ret[
	label = "{{Placeholder\lname : \"ret\"\llayout : *\loutput : float\<1 x 10 x 20 x 16\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

ret:Output -> ret_save:Output [dir="both"];
conv:Result -> ret_save:Input;
filter__2:Output -> conv:Filter;
bias__2:Output -> conv:Bias;
A:Output -> conv:Input;
predicate__1:Output -> conv:w;
}

Comment (SplitInfinity) [reply tests/unittests/GraphOptzTest.cpp]: Uh, can you grab a screenshot and post that? I think you can drag and drop the image into the text box.

Comment (mvsampath) [reply tests/unittests/GraphOptzTest.cpp]: optimizedF_
F_

File: tests/unittests/GraphOptzTest.cpp [hunk 5]
```diff
@@ -373,17 +373,21 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvWithPred) {
...
+  ASSERT_EQ(A->getNumUsers(), 2);
+  Node *newCV = A->getUsers().back().getUser();
   EXPECT_TRUE(llvm::isa<ConvolutionNode>(newCV));
   ASSERT_TRUE(newCV->hasPredicate());
   EXPECT_EQ(newCV->getPredicate().getNode(), pred2);
   ASSERT_EQ(newCV->getNumUsers(), 1);
   Node *save = newCV->getUsers().begin()->getUser();
   EXPECT_TRUE(llvm::isa<SaveNode>(save));
+
+  bindings_.allocate(A)->getHandle().randomize(-1.0, 1.0, mod_.getPRNG());
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]: What about the other `Placeholders` in the graph? Don't they need to be allocated too?

Comment (mvsampath) [reply tests/unittests/GraphOptzTest.cpp]: Fixed.

File: tests/unittests/GraphOptzTest.cpp [hunk 6]
```diff
@@ -373,17 +373,22 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvWithPred) {
...
+  ASSERT_EQ(A->getNumUsers(), 2);
+  Node *newCV = A->getUsers().back().getUser();
   EXPECT_TRUE(llvm::isa<ConvolutionNode>(newCV));
   ASSERT_TRUE(newCV->hasPredicate());
   EXPECT_EQ(newCV->getPredicate().getNode(), pred2);
   ASSERT_EQ(newCV->getNumUsers(), 1);
   Node *save = newCV->getUsers().begin()->getUser();
   EXPECT_TRUE(llvm::isa<SaveNode>(save));
+
+  bindings_.allocate(mod_.getPlaceholders());
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]: Okay, so the other `Placeholders` have been allocated but they need to be initialized too.

Comment (mvsampath) [reply tests/unittests/GraphOptzTest.cpp]: Updated. I am new to the code base and still trying to figure out how everything works.
Thanks for your patience.

**Comment by SplitInfinity** [comment]
@mvsampath Do you plan to land this?

**Comment by mvsampath** [comment]
> @mvsampath Do you plan to land this?

Yes. But I don't have permission to merge here. Let me check with someone and get back on how I can land this.
